docs/lang/articles/reference.md

```diff
@@ -259,22 +263,98 @@ stride) inside have to be evaluated to Python values.
+
+Calling a type annotation for literals
+
 ### The power operator
 
+```
+power ::= primary ["**" u_expr]
+```
+
+Applying '**' operator inside Taichi scope has the same behavior as Python. The power operator binds more tightly than unary operators on the left, but less tightly than unary operators on the right; i.e. -3 ** -2 == -(3 ** (-2)).
+
 ### Unary arithmetic and bitwise operations
 
+```
+u_expr ::= power | "-" power | "~" power
+```
+The unary - operator yields the negation of its argument. The unary ~ operator yields the bitwise inversion of its argument. - can be used with all scalar and tensor. ~ can only be used with integer scalar (i32, i64, etc.) and interrelated tensor.
+
 ### Binary arithmetic operations
 
+```
+m_expr ::= u_expr | m_expr "*" u_expr | m_expr "@" m_expr | m_expr "//" u_expr | m_expr "/" u_expr | m_expr "%" u_expr
+a_expr ::= m_expr | a_expr "+" m_expr | a_expr "-" m_expr
+```
+The binary arithmetic operators can operate on scalar and tensor. For tensor-tensor ops, both arguments must have the same shape. For scalar-tensor or tensor-scalar ops, the scalar is usually broadcast to the size of the tensor. The @ operator is for matrix multiplication and only operates on Tensor arguments.
+
+In division `/` operation, the compiler will automatically convert integral operands into default floating-point types (`f32` or `f64`).
+
+When the operands have different types, the outcoming type will be set as the expected type under C++ (e.g. `i32` + `i64` = `i64`).
+
 ### Shifting operations
 
+```
+shift_expr::= a_expr | shift_expr ( "<<" | ">>" ) a_expr
+```
+These operators accept integer scalar (i32, i64, etc.) and interrelated tensor for both arguments. When both arguments are tensors, they must have the same shape. When one is a scalar and the other is a tensor, the scalar is logically broadcast to match the size of the tensor.
+
+The compiler will check both operands to be integral type.
+
 ### Binary bitwise operations
 
+```
+and_expr ::= shift_expr | and_expr "&" shift_expr
+xor_expr ::= and_expr | xor_expr "^" and_expr
+or_expr  ::= xor_expr | or_expr "|" xor_expr
+```
+The & operator computes the bitwise AND of its arguments, the ^ the bitwise XOR, and the | the bitwise OR. The types requirements and broadcast logic are the same as shifting operations.
+
+The compiler will check both operands to be integral type.
+
 ### Comparisons
 
+```
+comparison    ::= or_expr (comp_operator or_expr)*
+comp_operator ::= "<" | ">" | "==" | ">=" | "<=" | "!=" | "is" ["not"] | ["not"] "in"
+```
+A comparison yields a boolean value (True or False), or if one of the operands is a Tensor, a boolean Tensor. Comparisons can be chained arbitrarily as long as they do not yield boolean Tensors that have more than one element. a op1 b op2 c ... is equivalent to a op1 b and b op2 c and ....
+
+All operands need to be primaries. The return type is `i32` for all comparision operations.
```

Comment:
Comparisons can be chained arbitrarily, e.g., `x < y <= z` is equivalent to `(x < y) & (y <= z)`.